bash_rsync.sh

```diff
@@ -2,40 +2,49 @@
 #
 # Name: bash_rsync.sh
 #
-# Brief: Bash script with user prompts so user does not forget what to enter as
-# source and destination, since it is a bit tricky. Otherwise, you can easily 
-# end up with extra directories you did not want. 
-# 
+# Brief: Bash script that uses rsync to backup files & folders from computer 
+# to external hard drive, USB drive, etc. 
+# - Script includes user prompts, since including or excluding a trailing / 
+# is tricky when entering the source and destination/target.
+# - Prevents user from ending up with extra unwanted sub-directories.
+#
 # Author: Kim Lew
 
-# Display user prompts of what to enter.
-echo "Type the source directory path & include specific directory\
-(/Users/kimlew/Documents):"
+# Display user prompts of what to enter to prevent creating extra sub-directories.
+# For the source: 
+# Enter the specific sub-directory WITHOUT /. rsync is smart enough to
+# create the sub-directory if it doesn't exist & transfers contents. 
+# If sub-directory already exists, rsync transfers contents. 
+# For the destination/target: 
+# You do NOT need to state specific sub-directory BUT you do need /.
+echo "Type source directory path WITH specific sub-directory & LEAVE OFF trailing /."
+echo "(Example: /Users/kimlew/Documents):"
```

Comment:
Just a suggestion, but you could avoid having to tell the user to leave off a trailing slash if you use those `bash` substitution operators to trim off trailing slashes. Here are a few different ways you could approach that, each a little more complex than the last.

First way: deal with single trailing slashes. You could remove a single trailing slash like this after you have read the user's value for `source_path`:

```
source_path=${source_path%/}
```

Second way: deal with consecutive trailing slashes. `bash` lets you put in consecutive slashes where a single slash would do. For example, these commands all print the same thing:

```
ls /Users/kimlew
ls //Users/kimlew
ls /////Users///kimlew////
```

so to be extra careful, you could remove consecutive trailing slashes at the end if you use the `shopt -s extglob` command to turn on "extended globbing" in `bash`, and then use a little more elaborate pattern with the `%%` operator:

```
# turn on bash extended globbing
shopt -s extglob

# remove all trailing slashes from the paths
source_path=${source_path%%+(/)}
destination_path=${destination_path%%+(/)}
```

The pattern `+(/)` means one or more slashes, and the operator `%%` means remove the longest matching string at the end (see https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html, in particular the section on the `%%` operator, and the linked documentation about 'Pattern Matching' for details).

Third way: handle the case where one of the paths is the root directory. If you want to be extra, extra careful, you should check whether the source or destination paths are the root directory. Since the root directory is signified by just a slash `/`, if you remove all the trailing slashes from those paths you'll end up with an empty string, which isn't what you want. A way to safeguard against this is to use a test and the `||` operator:

```
# turn on bash extended globbing
shopt -s extglob

# remove all trailing slashes from the paths
[ "$source_path" == '/' ] || source_path=${source_path%%+(/)}
[ "$destination_path" == '/' ] || destination_path=${destination_path%%+(/)}
```

Fourth way: even better handling if the path is the root directory. Being even more careful, if you're going to handle multiple slashes at the end, you should also handle the case where one of the paths is `//`, `///`, etc. i.e. where the path is the root directory, but it's given with redundant slashes.

Probably the easiest way to do this is to modify the paths only if they aren't empty, and then check if they are empty after you modified them. If they are, it means the original path was just a series of slashes, so it's meant to be the root directory:

```
# turn on bash extended globbing
shopt -s extglob

# remove all trailing slashes from the paths
if [ "$source_path" != '' ]; then
  source_path=${source_path%%+(/)}
  if [ "$source_path" == '' ]; then
    source_path=/
  fi
fi

# ... and do the same for the destination_path
```

That's getting a little wordy, so you could capture the slash removal in a function:

```
# turn on bash extended globbing
shopt -s extglob

function remove_trailing_slashes() {
  local path=$1
  if [ "$path" != '' ]; then
    path=${path%%+(/)}
    if [ "$path" == '' ]; then
      path=/
    fi
  fi
  echo "$path"
}

source_path=$(remove_trailing_slashes "$source_path")
destination_path=$(remove_trailing_slashes "$destination_path")
```

Note that `remove_trailing_slashes` echos the new value of `path` -- that's the only way to 'return' it to the caller. In the assignments to `source_path` and `destination_path`, we're using `$(...)` with ordinary parentheses, not curly braces, to capture the standard output of the call to `remove_trailing_slashes`.

Reply:
@kcharter I will keep this in my notes for use in a different script, but am going to leave as it is for this one, since it is for a very specific singular purpose.

Reply:
OK, sounds good @kimlew